File: scripts/spool_tag.py
#!/usr/bin/env python3
"""
Generate raw NFC page data for Anycubic ACE Pro compatible filament tags.

No reader hardware or web server required. This prints a hex blob you can
paste into a phone NFC app (e.g. NFC Tools Pro -> Other -> Write raw data)
along with the start page to write it at.

Format reverse-engineered by https://github.com/Molodos/anycubic-nfc-filament
(see format.md in that repo). Tag is an NTAG213: pages 0x00-0x03 and
0x28-0x2c are chip/management data and must never be touched. Pages
0x04-0x27 (36 pages / 144 bytes) hold the filament spec and are what this
tool produces.
"""

import argparse
import sys

START_PAGE = 0x04
END_PAGE = 0x27  # inclusive
PAGE_COUNT = END_PAGE - START_PAGE + 1

SKUS = {
    "PLA": "AHPLBK-101",
    "PLA+": "AHPLPBK-102",
    "PLA High Speed": "AHHSBK-102",
    "PLA Matte": "HYGBK-101",
    "PLA Silk": "HSCWH-101",
    "PETG": "HPEBK-103",
    "ASA": "HASBK-101",
    "ABS": "HABBK-102",
    "TPU": "HTPBK-101",
    "PLA Luminous": "HFGBL-101",
}


class TagPages:
    def __init__(self):
        self.pages = [bytearray(4) for _ in range(PAGE_COUNT)]

    def _rel(self, page):
        idx = page - START_PAGE
        if not (0 <= idx < PAGE_COUNT):
            raise ValueError(f"page {page:#x} outside writable range 0x04-0x27")
        return idx

    def write_byte(self, page, index, value):
        self.pages[self._rel(page)][index] = value & 0xFF

    def write_u16(self, page, index, value):
        low, high = value % 256, value // 256
        self.write_byte(page, index, low)
        if index == 3:
            self.write_byte(page + 1, 0, high)
        else:
            self.write_byte(page, index + 1, high)

    def write_string(self, page, text, max_len=20):
        for i in range(min(len(text), max_len)):
            self.write_byte(page + i // 4, i % 4, ord(text[i]))

    def write_color(self, page, hex_color):
        hex_color = hex_color.lstrip("#")
        r, g, b = (int(hex_color[i:i + 2], 16) for i in (0, 2, 4))
        self.write_byte(page, 0, 0xFF)  # alpha/opacity
        self.write_byte(page, 1, b)
        self.write_byte(page, 2, g)
        self.write_byte(page, 3, r)

    def hex_blob(self):
        return "".join(page.hex() for page in self.pages)

    def page_dump(self):
        lines = []
        for i, page in enumerate(self.pages):
            lines.append(f"[Page {START_PAGE + i:02x}] {page.hex(':')}")
        return "\n".join(lines)


def build_tag(spec):
    t = TagPages()

    # Static markers
    t.write_byte(0x04, 0, 0x7B)
    t.write_byte(0x04, 2, 0x65)  # format version 2
    t.write_byte(0x27, 3, 0x4D)  # custom spool marker

    t.write_string(0x05, SKUS.get(spec["type"], "AHPLBK-101"))
    t.write_string(0x0A, spec.get("manufacturer", "AC"))
    t.write_string(0x0F, spec["type"])

    t.write_color(0x14, spec["color"])

    t.write_u16(0x17, 0, spec.get("speed_min", 0))
    t.write_u16(0x17, 2, spec.get("speed_max", 0))
    t.write_u16(0x18, 0, spec["nozzle_min"])
    t.write_u16(0x18, 2, spec["nozzle_max"])

    t.write_u16(0x1D, 0, spec["bed_min"])
    t.write_u16(0x1D, 2, spec["bed_max"])

    t.write_u16(0x1E, 0, round(spec["diameter"] * 100))
    t.write_u16(0x1E, 2, spec["length"])
    t.write_u16(0x1F, 0, spec["weight"])

    return t


def prompt(text, default=None, cast=str):
    suffix = f" [{default}]" if default is not None else ""
    while True:
        raw = input(f"{text}{suffix}: ").strip()
        if not raw and default is not None:
            return default
        if not raw:
            print("  A value is required.")
            continue
        try:
            return cast(raw)
        except ValueError:
            print("  Invalid value, try again.")


def interactive_spec():
    print("Anycubic NFC spool tag generator\n")

    types = list(SKUS.keys())
    print("Filament types:")
    for i, tname in enumerate(types, 1):
        print(f"  {i}. {tname}")
    choice = prompt("Type number", default="1", cast=int)
    ftype = types[choice - 1] if 1 <= choice <= len(types) else types[0]

    manufacturer = prompt("Manufacturer code (2-4 letters)", default="AC")
    color = prompt("Color hex (e.g. #89a84f)", default="#ffffff")

    nozzle_min = prompt("Nozzle temp min (C)", default=200, cast=int)
    nozzle_max = prompt("Nozzle temp max (C)", default=210, cast=int)
    bed_min = prompt("Bed temp min (C)", default=50, cast=int)
    bed_max = prompt("Bed temp max (C)", default=60, cast=int)
    speed_min = prompt("Print speed min mm/s (0 = skip)", default=0, cast=int)
    speed_max = prompt("Print speed max mm/s (0 = skip)", default=0, cast=int)

    diameter = prompt("Filament diameter mm", default=1.75, cast=float)
    length = prompt("Spool length (m)", default=330, cast=int)
    weight = prompt("Spool weight (g)", default=1000, cast=int)

    return {
        "type": ftype,
        "manufacturer": manufacturer,
        "color": color,
        "nozzle_min": nozzle_min,
        "nozzle_max": nozzle_max,
        "bed_min": bed_min,
        "bed_max": bed_max,
        "speed_min": speed_min,
        "speed_max": speed_max,
        "diameter": diameter,
        "length": length,
        "weight": weight,
    }


def cli_spec(args):
    return {
        "type": args.type,
        "manufacturer": args.manufacturer,
        "color": args.color,
        "nozzle_min": args.nozzle_min,
        "nozzle_max": args.nozzle_max,
        "bed_min": args.bed_min,
        "bed_max": args.bed_max,
        "speed_min": args.speed_min,
        "speed_max": args.speed_max,
        "diameter": args.diameter,
        "length": args.length,
        "weight": args.weight,
    }


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--type", choices=list(SKUS.keys()))
    parser.add_argument("--manufacturer", default="AC")
    parser.add_argument("--color", help="hex color, e.g. #89a84f")
    parser.add_argument("--nozzle-min", type=int, dest="nozzle_min")
    parser.add_argument("--nozzle-max", type=int, dest="nozzle_max")
    parser.add_argument("--bed-min", type=int, dest="bed_min")
    parser.add_argument("--bed-max", type=int, dest="bed_max")
    parser.add_argument("--speed-min", type=int, dest="speed_min", default=0)
    parser.add_argument("--speed-max", type=int, dest="speed_max", default=0)
    parser.add_argument("--diameter", type=float, default=1.75)
    parser.add_argument("--length", type=int, help="spool length in meters")
    parser.add_argument("--weight", type=int, help="spool weight in grams")
    parser.add_argument("--dump", action="store_true",
                         help="also print a human-readable per-page dump")
    args = parser.parse_args()

    if args.type and args.color and args.nozzle_min is not None:
        spec = cli_spec(args)
    else:
        spec = interactive_spec()

    tag = build_tag(spec)
    blob = tag.hex_blob()

    print("\n--- Copy into NFC Tools Pro: Other -> Write raw data ---")
    print(f"Start page: {START_PAGE} (0x{START_PAGE:02x})")
    print(f"Byte count: {len(blob) // 2} ({PAGE_COUNT} pages)")
    print("Hex data:")
    print(blob)

    if args.dump:
        print("\n--- Page dump ---")
        print(tag.page_dump())

    print(
        "\nReminder: each spool needs two tags written identically. "
        "Never write to pages 0x00-0x03 or 0x28-0x2c."
    )


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        sys.exit(1)

File: app/src/main/java/com/jamieduncan/acetag/SpoolTag.kt
package com.jamieduncan.acetag

/**
 * Encodes filament spool specs into the Anycubic ACE Pro NFC tag format.
 * Format reverse-engineered by https://github.com/Molodos/anycubic-nfc-filament
 * (see format.md). Layout is NTAG213: pages 0x00-0x03 and 0x28-0x2c are
 * chip/management data and must never be written. Pages 0x04-0x27 (36 pages,
 * 4 bytes each) hold the filament spec produced here.
 */
object SpoolTag {
    const val START_PAGE = 0x04
    const val END_PAGE = 0x27 // inclusive
    const val PAGE_COUNT = END_PAGE - START_PAGE + 1

    // Page 0x20 is always zero in every known Anycubic dump (both v1 and v2), same as the
    // 0x27-byte-3 "custom spool marker" this format already repurposes. We use it to write a
    // random 4-byte group ID, shared by both tags of one spool, so a future read can pair them
    // deterministically instead of guessing by matching specs (which breaks for two identical
    // -color spools). Absent/zero on genuine Anycubic tags and tags written before this existed.
    const val GROUP_ID_PAGE = 0x20

    // Byte 3 of the last page is 0x4D on every tag this app writes and 0x00 on every genuine
    // Anycubic dump. It's what lets a scan tell "a spool I've never seen" from "a sticker I wrote
    // that isn't on a spool any more" — the second is offered to an existing spool rather than
    // added as a new one, which is the whole point of tags being movable.
    const val CUSTOM_MARKER_PAGE = 0x27
    private const val CUSTOM_MARKER_INDEX = 3
    private const val CUSTOM_MARKER = 0x4D

    fun randomGroupId(): ByteArray = ByteArray(4).also { kotlin.random.Random.nextBytes(it) }

    data class Defaults(
        val nozzleMin: Int,
        val nozzleMax: Int,
        val bedMin: Int,
        val bedMax: Int,
        val speedMin: Int = 0,
        val speedMax: Int = 0,
        val diameterMm: Double = 1.75,
        val lengthM: Int = 330,
        val weightG: Int = 1000,
    )

    // Typical published ranges per material; a starting point, always user-editable. Keyed by tag
    // type string, so the form falls back to the base material's numbers for any finish that
    // doesn't have its own entry.
    val MATERIAL_DEFAULTS: Map<String, Defaults> = mapOf(
        "PLA" to Defaults(nozzleMin = 190, nozzleMax = 220, bedMin = 45, bedMax = 60),
        "PLA+" to Defaults(nozzleMin = 200, nozzleMax = 225, bedMin = 45, bedMax = 60),
        "PLA High Speed" to Defaults(nozzleMin = 190, nozzleMax = 240, bedMin = 45, bedMax = 60, speedMin = 50, speedMax = 300),
        "PLA Matte" to Defaults(nozzleMin = 190, nozzleMax = 220, bedMin = 45, bedMax = 60),
        "PLA Silk" to Defaults(nozzleMin = 200, nozzleMax = 220, bedMin = 45, bedMax = 60),
        "PETG" to Defaults(nozzleMin = 230, nozzleMax = 250, bedMin = 70, bedMax = 90),
        "ASA" to Defaults(nozzleMin = 240, nozzleMax = 260, bedMin = 90, bedMax = 100),
        "ABS" to Defaults(nozzleMin = 230, nozzleMax = 250, bedMin = 90, bedMax = 100),
        "TPU" to Defaults(nozzleMin = 210, nozzleMax = 230, bedMin = 30, bedMax = 60),
        "PLA Luminous" to Defaults(nozzleMin = 190, nozzleMax = 220, bedMin = 45, bedMax = 60),
    )

    data class Spec(
        val type: String,
        val manufacturer: String,
        val color: String,
        val nozzleMin: Int,
        val nozzleMax: Int,
        val bedMin: Int,
        val bedMax: Int,
        val speedMin: Int,
        val speedMax: Int,
        val diameterMm: Double,
        val lengthM: Int,
        val weightG: Int,
    )

    /** 36 pages of 4 bytes each, indices correspond to pages START_PAGE..END_PAGE. */
    class Pages {
        val pages: Array<ByteArray> = Array(PAGE_COUNT) { ByteArray(4) }

        private fun rel(page: Int): Int {
            val idx = page - START_PAGE
            require(idx in 0 until PAGE_COUNT) { "page 0x${page.toString(16)} outside writable range" }
            return idx
        }

        fun writeByte(page: Int, index: Int, value: Int) {
            pages[rel(page)][index] = (value and 0xFF).toByte()
        }

        fun writeU16(page: Int, index: Int, value: Int) {
            val low = value % 256
            val high = value / 256
            writeByte(page, index, low)
            if (index == 3) {
                writeByte(page + 1, 0, high)
            } else {
                writeByte(page, index + 1, high)
            }
        }

        fun writeString(page: Int, text: String, maxLen: Int = 20) {
            val n = minOf(text.length, maxLen)
            for (i in 0 until n) {
                writeByte(page + i / 4, i % 4, text[i].code)
            }
        }

        fun writeColor(page: Int, hexColor: String) {
            val h = hexColor.removePrefix("#")
            val r = h.substring(0, 2).toInt(16)
            val g = h.substring(2, 4).toInt(16)
            val b = h.substring(4, 6).toInt(16)
            writeByte(page, 0, 0xFF) // alpha/opacity
            writeByte(page, 1, b)
            writeByte(page, 2, g)
            writeByte(page, 3, r)
        }

        fun readByte(page: Int, index: Int): Int {
            return pages[rel(page)][index].toInt() and 0xFF
        }

        fun readU16(page: Int, index: Int): Int {
            val low = readByte(page, index)
            val high = if (index == 3) readByte(page + 1, 0) else readByte(page, index + 1)
            return high * 256 + low
        }

        fun readString(page: Int, maxLen: Int = 20): String {
            val sb = StringBuilder()
            var p = page
            var i = 0
            var b = readByte(p, i)
            while (b > 0) {
                sb.append(b.toChar())
                if (sb.length >= maxLen) break
                i++
                if (i == 4) {
                    i = 0
                    p++
                }
                b = readByte(p, i)
            }
            return sb.toString()
        }

        fun readColor(page: Int): String {
            val a = readByte(page, 0)
            val b = readByte(page, 1)
            val g = readByte(page, 2)
            val r = readByte(page, 3)
            if (a == 0 && r == 0 && g == 0 && b == 0) return ""
            return String.format("#%02x%02x%02x", r, g, b)
        }

        fun writeGroupId(bytes: ByteArray) {
            require(bytes.size == 4) { "group id must be exactly 4 bytes" }
            for (i in 0..3) writeByte(GROUP_ID_PAGE, i, bytes[i].toInt() and 0xFF)
        }

        /** Null if the tag has no group ID written (genuine Anycubic tags, or pre-group-ID writes). */
        fun readGroupIdHex(): String? {
            val bytes = ByteArray(4) { readByte(GROUP_ID_PAGE, it).toByte() }
            if (bytes.all { it == 0.toByte() }) return null
            return bytes.joinToString("") { "%02x".format(it) }
        }

        companion object {
            /** Builds Pages from 144 raw bytes read starting at START_PAGE (as returned by MifareUltralight.readPages). */
            fun fromBytes(bytes: ByteArray): Pages {
                require(bytes.size == PAGE_COUNT * 4) { "expected ${PAGE_COUNT * 4} bytes, got ${bytes.size}" }
                val t = Pages()
                for (i in 0 until PAGE_COUNT) {
                    for (j in 0..3) {
                        t.pages[i][j] = bytes[i * 4 + j]
                    }
                }
                return t
            }
        }
    }

    /** True if this app wrote the tag, rather than it coming off a factory Anycubic spool. */
    fun isCustomWritten(t: Pages): Boolean =
        t.readByte(CUSTOM_MARKER_PAGE, CUSTOM_MARKER_INDEX) == CUSTOM_MARKER

    /**
     * How much filament a spool of [weightG] holds, scaled from the material's own full-spool
     * figures. A 250 g spool written with a 1 kg length tells the ACE it has four times the
     * filament it does, so the remaining-length estimate is wrong from the first gram.
     *
     * Zero when the material has no published defaults to scale from — the caller leaves the
     * field alone rather than inventing a number, same as [MATERIAL_DEFAULTS] lookups elsewhere.
     */
    fun lengthForWeight(type: String, weightG: Int): Int {
        val d = MATERIAL_DEFAULTS[type] ?: return 0
        if (weightG <= 0 || d.weightG <= 0) return 0
        return Math.round(d.lengthM.toDouble() * weightG / d.weightG).toInt()
    }

    /** Returns null if this doesn't look like an Anycubic-format tag (missing the format marker). */
    fun decode(t: Pages): Spec? {
        if (t.readByte(0x04, 0) != 0x7B) return null
        val type = t.readString(0x0F).ifBlank { t.readString(0x05) }
        return Spec(
            type = type,
            manufacturer = t.readString(0x0A),
            color = t.readColor(0x14),
            speedMin = t.readU16(0x17, 0),
            speedMax = t.readU16(0x17, 2),
            nozzleMin = t.readU16(0x18, 0),
            nozzleMax = t.readU16(0x18, 2),
            bedMin = t.readU16(0x1D, 0),
            bedMax = t.readU16(0x1D, 2),
            diameterMm = t.readU16(0x1E, 0) / 100.0,
            lengthM = t.readU16(0x1E, 2),
            weightG = t.readU16(0x1F, 0),
        )
    }

    fun buildTag(spec: Spec, groupId: ByteArray? = null): Pages {
        val t = Pages()

        // Static markers
        t.writeByte(0x04, 0, 0x7B)
        t.writeByte(0x04, 2, 0x65) // format version 2
        t.writeByte(CUSTOM_MARKER_PAGE, CUSTOM_MARKER_INDEX, CUSTOM_MARKER)
        groupId?.let { t.writeGroupId(it) }

        t.writeString(0x05, FilamentMaterial.skuForTagType(spec.type))
        t.writeString(0x0A, spec.manufacturer.ifBlank { "AC" })
        t.writeString(0x0F, spec.type)

        t.writeColor(0x14, spec.color)

        t.writeU16(0x17, 0, spec.speedMin)
        t.writeU16(0x17, 2, spec.speedMax)
        t.writeU16(0x18, 0, spec.nozzleMin)
        t.writeU16(0x18, 2, spec.nozzleMax)

        t.writeU16(0x1D, 0, spec.bedMin)
        t.writeU16(0x1D, 2, spec.bedMax)

        t.writeU16(0x1E, 0, Math.round(spec.diameterMm * 100).toInt())
        t.writeU16(0x1E, 2, spec.lengthM)
        t.writeU16(0x1F, 0, spec.weightG)

        return t
    }
}
